Trait upcasting coercion (part2)

crlf0710 opened the pull request:
This is the second part of trait upcasting coercion implementation.

Currently this is blocked on #86264 .

The third part might be implemented using unsafety checking

r? @bjorn3

rust-highfive commented:
Some changes occured to rustc_codegen_cranelift

cc @bjorn3

Some changes occured to the CTFE / Miri engine

cc @rust-lang/miri

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3172985008) [(plain)](https://github.com/rust-lang/rust/commit/522611b6b854f7319f128ccb8f271d2709d8889f/checks/3172985008/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......................................................................ii........................... 12100/12180
................................................................................
failures:

---- [ui] ui/traits/trait-upcasting/replace-vptr.rs stdout ----

error: test compilation failed although it shouldn't!
status: signal: 11 (core dumped)
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/traits/trait-upcasting/replace-vptr.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/trait-upcasting/replace-vptr/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/trait-upcasting/replace-vptr/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x60b793)[0x7fdcc65fa793]
/lib/x86_64-linux-gnu/libc.so.6(+0x3f040)[0x7fdcc58b3040]
/usr/lib/x86_64-linux-gnu/libLLVM-10.so.1(_ZN4llvm11PointerType3getEPNS_4TypeEj+0x12)[0x7fdcc1c3eda2]
/usr/lib/x86_64-linux-gnu/libLLVM-10.so.1(+0x9cef5a)[0x7fdcc1b50f5a]
/usr/lib/x86_64-linux-gnu/libLLVM-10.so.1(+0x9ee66c)[0x7fdcc1b7066c]
/usr/lib/x86_64-linux-gnu/libLLVM-10.so.1(+0xa00be4)[0x7fdcc1b82be4]
/usr/lib/x86_64-linux-gnu/libLLVM-10.so.1(LLVMBuildStructGEP+0x49)[0x7fdcc1b80669]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x86ffd6)[0x7fdcc685efd6]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x934a1d)[0x7fdcc6923a1d]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x927e3e)[0x7fdcc6916e3e]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x872e16)[0x7fdcc6861e16]
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x8b8aa3)[0x7fdcc68a7aa3]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x904155)[0x7fdcc68f3155]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x839ea1)[0x7fdcc6828ea1]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x903bb2)[0x7fdcc68f2bb2]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x871e11)[0x7fdcc6860e11]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(_ZN109_$LT$rustc_codegen_llvm..LlvmCodegenBackend$u20$as$u20$rustc_codegen_ssa..traits..backend..CodegenBackend$GT$13codegen_crate17h69d6935174cd2c1aE+0x69)[0x7fdcc689cdd9]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x7add01)[0x7fdcc679cd01]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x7462ba)[0x7fdcc67352ba]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(_ZN15rustc_interface7queries7Queries15ongoing_codegen17h5aa42a039abe4eb4E+0xb8)[0x7fdcc6728988]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x65266b)[0x7fdcc664166b]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x60dbbc)[0x7fdcc65fcbbc]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x653e7c)[0x7fdcc6642e7c]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x649ea3)[0x7fdcc6638ea3]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x60fa2f)[0x7fdcc65fea2f]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x67a496)[0x7fdcc6669496]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/librustc_driver-4c5decbaebf19c76.so(+0x6081ea)[0x7fdcc65f71ea]
/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/libstd-acc102086c90e281.so(rust_metadata_std_21736b398a6b3262+0xaafaa)[0x7fdcc5d0ffaa]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76db)[0x7fdcc09c96db]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7fdcc599571f]
------------------------------------------




failures:
    [ui] ui/traits/trait-upcasting/replace-vptr.rs
test result: FAILED. 12076 passed; 1 failed; 103 ignored; 0 measured; 0 filtered out; finished in 127.76s




command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:09
```

</details>

bjorn3 commented on compiler/rustc_codegen_cranelift/src/unsize.rs:
```diff
@@ -58,24 +86,29 @@ fn unsize_thin_ptr<'tcx>(
         (&ty::Adt(def_a, _), &ty::Adt(def_b, _)) => {
```
Above `unsized_info` is getting `None` passed as `old_info` twice. I think these need to be passed the real `old_info`.

crlf0710 replied:
Currently typechecking will ensure `old_info` is `None` in these cases. But it's no harm and more consistent to always pass `old_info` inside. I'll do this.

crlf0710 replied:
Done.

bjorn3 commented on compiler/rustc_codegen_cranelift/src/unsize.rs:
```diff
@@ -58,24 +86,29 @@ fn unsize_thin_ptr<'tcx>(
-                    continue;
+            if src_layout == dst_layout {
+                assert!(old_info.is_some());
+                (src, old_info.unwrap())
```
Can you use a `return` here to avoid indenting the rest of the code?

crlf0710 replied:
Done

bjorn3 commented on compiler/rustc_codegen_cranelift/src/unsize.rs:
```diff
@@ -58,24 +86,29 @@ fn unsize_thin_ptr<'tcx>(
-                if src_f.is_zst() {
-                    continue;
+            if src_layout == dst_layout {
+                assert!(old_info.is_some());
```
```suggestion
```
You are already unwrapping it below.

crlf0710 replied:
Done

bjorn3 commented on compiler/rustc_codegen_cranelift/src/unsize.rs:
```diff
@@ -93,10 +126,11 @@ pub(crate) fn coerce_unsized_into<'tcx>(
             if fx.layout_of(src.layout().ty.builtin_deref(true).unwrap().ty).is_unsized() {
                 // fat-ptr to fat-ptr unsize preserves the vtable
                 // i.e., &'a fmt::Debug+Send => &'a fmt::Debug
```
```suggestion
```
outdated

crlf0710 replied:
Done

bjorn3 commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -128,69 +128,107 @@ pub fn compare_simd_types<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
-                    continue;
+            if src_ty == dst_ty {
+                assert!(old_info.is_some());
+                result = Some((src, old_info.unwrap()));
```
Same early return here please.

crlf0710 replied:
Done.

RalfJung commented on compiler/rustc_mir/src/interpret/cast.rs:
```diff
@@ -269,12 +269,61 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                    if let Some(entry_idx) = vptr_entry_idx {
+                        let (old_data, old_vptr) = match *val {
+                            Immediate::ScalarPair(data, vptr) => (data, vptr),
+                            Immediate::Scalar(..) => span_bug!(
```
Please don't `match` on `Immediate` here; instead add a method like `to_scalar_or_uninit_pair` or so to `Immediate` in `operand.rs`.

I am a bit surprised we don't have such a method yet; it's probably because so far all interactions with pointers to unsized types went through `MemPlace` (which has a `vtable()` method) -- but creating a `MemPlace` here would make little sense.

RalfJung replied:
In fact you can make it `to_scalar_pair` since you are doing `check_init` on both components anyway.

(This matches the existing `to_scalar_or_uninit` and `to_scalar` methods.)

crlf0710 replied:
Done.

RalfJung commented on compiler/rustc_mir/src/interpret/cast.rs:
```diff
@@ -269,12 +269,61 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                        let entry_idx = u64::try_from(entry_idx).unwrap();
+                        let old_data = old_data.check_init()?;
+                        let old_vptr = self.scalar_to_ptr(old_vptr.check_init()?);
+                        let vtable = self
```
Please don't interact with the vtable directly in this file. Instead, add a new method in `traits.rs` (similar to how we have `get_vtable_slot` to read the fn ptr at a particular slot from the vtable).

crlf0710 replied:
Done.

bors commented:
:umbrella: The latest upstream changes (presumably #83484) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

crlf0710 commented:
#86264 has landed, now this is unblocked. I'll address the review comments very soon, and add more tests.

crlf0710 commented:
Rebased, addressed the review comments and pull in some tests and document from previous work. This is ready for review again now.

bjorn3 commented on compiler/rustc_codegen_cranelift/src/unsize.rs:
```diff
@@ -25,39 +25,70 @@ pub(crate) fn unsized_info<'tcx>(
+                .expect("unsized_info: missing principal trait for trait upcasting coercion");
+            let principal_b = data_b
+                .principal()
+                .expect("unsized_info: missing principal trait for trait upcasting coercion");
```
Isn't the principal non-existent when the trait object only consists of traits like `Send`, `Sync` or `Sized`?

crlf0710 replied:
Yes, if the principal is non-existent, either it's not upcasting coercion, like `&dyn Send+Sync` to `&dyn Send` will be a simple case (and will take the early return code path), or it will be rejected by type-checking, like `dyn Foo + Send` to `dyn Send` will be rejected.

I added some comments within the code to clarify.

bjorn3 commented on compiler/rustc_codegen_cranelift/src/unsize.rs:
```diff
@@ -25,39 +25,70 @@ pub(crate) fn unsized_info<'tcx>(
+                let vptr_ptr = Pointer::new(old_info).offset_i64(fx, entry_offset.into()).load(
+                    fx,
+                    fx.pointer_type,
+                    MemFlags::trusted(),
```
Can you please use `crate::vtable::vtable_memflags()` here instead? (You need to `pub(crate)` it first.) `vtable_memflags()` sets the readonly flag that specifies that the loaded memory won't be changed for the duration of the call. cg_ssa does the same with `set_invariant_load`.

crlf0710 replied:
Done!

bjorn3 commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -218,16 +252,13 @@ pub fn coerce_unsized_into<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+                    let base = bx.bitcast(
+                        base,
+                        bx.cx().scalar_pair_element_backend_type(dst.layout, 0, true),
+                    );
```
Shouldn't this bitcast happen inside `unsize_ptr` if necessary?

crlf0710 replied:
Indeed! Thanks!

crlf0710 commented:
Thanks for the comments! Updated.

RalfJung commented on compiler/rustc_mir/src/interpret/traits.rs:
```diff
@@ -121,4 +121,27 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                pointer_size * idx.checked_add(1).unwrap(),
+                self.tcx.data_layout.pointer_align.abi,
+            )?
+            .expect("cannot be a ZST");
```
For consistency, can you write this code like `get_vtable_slot`?
Currently you are conceptually reading the first `idx` entries of the vtable, but its more accurate IMO to just read the one you actually need.

So, something like
```rust
let vtable_slot = vtable.offset(ptr_size * idx, self)?;
let vtable_slot = self
            .memory
            .get(vtable_slot, ptr_size, self.tcx.data_layout.pointer_align.abi)?
.expect("cannot be a ZST");
```

crlf0710 replied:
Sure. Done.

RalfJung commented on compiler/rustc_mir/src/interpret/traits.rs:
```diff
@@ -121,4 +121,27 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+        let new_vtable = self
+            .scalar_to_ptr(vtable.read_ptr_sized(pointer_size * idx)?.check_init()?)
+            .into_pointer_or_addr()
+            .expect("should be a pointer");
```
Why do you need a `Pointer<M::PointerTag>` here? I'd expect that this should return a `Pointer<Option<M::PointerTag>>`.

RalfJung replied:
Oh I see, `new_dyn_trait` requires a `Pointer<Tag>`.
That should not be needed, I think you can change that to `Pointer<Option<Tag>>`. (I can make a separate PR for that if you prefer.)

crlf0710 replied:
I've changed it now... Not sure if this is exactly what you want.

RalfJung commented on compiler/rustc_mir/src/interpret/operand.rs:
```diff
@@ -79,6 +79,16 @@ impl<'tcx, Tag: Provenance> Immediate<Tag> {
+        match self {
+            Immediate::ScalarPair(val1, val2) => Ok((val1.check_init()?, val2.check_init()?)),
+            Immediate::Scalar(..) => {
+                bug!("Got a scalar where a wide pointer was expected")
```
```suggestion
                bug!("Got a scalar where a scalar pair was expected")
```
Wide pointers are not the only scalar pairs.

crlf0710 replied:
Done. Thanks!

crlf0710 commented:
@RalfJung Thanks for the review! I've addressed the comments.

RalfJung commented on compiler/rustc_mir/src/interpret/operand.rs:
```diff
@@ -63,22 +63,42 @@ impl<'tcx, Tag: Provenance> Immediate<Tag> {
+    ) -> Self {
+        Immediate::ScalarPair(
+            val.into(),
+            ScalarMaybeUninit::from_pointer(
```
Please use `from_maybe_pointer`, then you don't need to `expect`.

crlf0710 replied:
Done. Thanks!

bjorn3 commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -128,55 +128,97 @@ pub fn compare_simd_types<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+            let lldata_ty = bx.cx().scalar_pair_element_backend_type(dst_layout, 0, true);
+            let llextra_ty = bx.cx().scalar_pair_element_backend_type(dst_layout, 1, true);
+            if src_ty == dst_ty {
+                let (lldata, llextra) = (src, old_info.unwrap());
+                // HACK(eddyb) have to bitcast pointers until LLVM removes pointee types.
+                return (bx.bitcast(lldata, lldata_ty), bx.bitcast(llextra, llextra_ty));
```
If the source and destination type are identical, you don't need a bitcast, right?

crlf0710 replied:
Indeed! Thanks.

bjorn3 reviewed (approved):
The codegen part LGTM apart from a minor nit. I will leave @RalfJung to approve the miri part. I don't see any obvious problems with the trait selection code, but given that I am not familiar with it, maybe @nikomatsakis want to take a look?

crlf0710 commented:
Updated and addressed the comments. 

For the trait selection related stuff, I think niko has been and will be on vacation in the next two weeks. if it's possible i'd prefer more detailed review of this part be defered a little, and we'll revisit it after [this FIXME](https://github.com/rust-lang/rust/blob/f99a37e08a33da21c8d9ce8b2d1523abe5d5186b/compiler/rustc_trait_selection/src/traits/mod.rs#L768) is solved.

RalfJung commented:
:+1: for the Miri engine changes.

bjorn3 commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -128,55 +128,95 @@ pub fn compare_simd_types<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+            let llextra_ty = bx.cx().scalar_pair_element_backend_type(dst_layout, 1, true);
+            if src_ty == dst_ty {
+                return (src, old_info.unwrap());
+            }
```
One last nit: You could do the `scalar_pair_element_backend_type` after this early return.

crlf0710 replied:
Done~

bjorn3 reviewed (approved):
> For the trait selection related stuff, I think niko has been and will be on vacation in the next two weeks. if it's possible i'd prefer more detailed review of this part be defered a little, and we'll revisit it after this FIXME is solved.

Sure. r=me with the nit resolved.

crlf0710 commented:
@bjorn3 Updated with the comment above addressed!

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 3cb625e4683c43991de6fc27d6b4e0db5a34011f has been approved by `bjorn3`

<!-- @bors r=bjorn3 3cb625e4683c43991de6fc27d6b4e0db5a34011f -->
<!-- homu: {"type":"Approved","sha":"3cb625e4683c43991de6fc27d6b4e0db5a34011f","approver":"bjorn3"} -->

bors commented:
:hourglass: Testing commit 3cb625e4683c43991de6fc27d6b4e0db5a34011f with merge c6bc102fea9f9274202d0bc55a0fef8a3bc92426...
<!-- homu: {"type":"BuildStarted","head_sha":"3cb625e4683c43991de6fc27d6b4e0db5a34011f","merge_sha":"c6bc102fea9f9274202d0bc55a0fef8a3bc92426"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3234606961)
Approved by: bjorn3
Pushing c6bc102fea9f9274202d0bc55a0fef8a3bc92426 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3234606961"},"merge_sha":"c6bc102fea9f9274202d0bc55a0fef8a3bc92426"} -->
